File: tests/fixtures/zig/enums.zig
const State = enum(i8) {
    negative = -3,
    zero = 0,
    ready = 7,
};

const Tagged = union(enum) {
    unit,
    integer: u32,
};

const Raw = extern union {
    integer: i32,
    floating: f32,
};

const Failure = error{bad}!u32;

noinline fn inspectEnums(
    state: *const State,
    tagged: *const Tagged,
    raw: *const Raw,
    optional: *const ?u32,
    failure: *const Failure,
) bool {
    asm volatile ("" :: [state] "r" (state), [tagged] "r" (tagged), [raw] "r" (raw), [optional] "r" (optional), [failure] "r" (failure) : .{ .memory = true });
    return state.* == .negative and tagged.integer == 42 and raw.integer == 42 and optional.*.? == 43 and failure.* catch 0 == 44;
}

pub fn main() !void {
    const state: State = .negative;
    const tagged: Tagged = .{ .integer = 42 };
    const raw: Raw = .{ .integer = 42 };
    const optional: ?u32 = 43;
    const failure: Failure = 44;
    if (!inspectEnums(&state, &tagged, &raw, &optional, &failure)) {
        return error.UnexpectedEnumFixtureValue;
    }
}

File: tests/fixtures/cpp/enums.cpp
#include <cstdint>

enum class State : std::int8_t {
    Negative = -3,
    Zero = 0,
    Alias = 0,
    Ready = 7,
};

union Raw {
    std::int32_t integer;
    float floating;
};

struct ManualTagged {
    enum class Kind : std::uint8_t {
        Integer,
        Floating,
    } kind;
    union {
        std::int32_t integer;
        float floating;
    } payload;
};

__attribute__((noinline)) static bool inspect_enums(const State* state,
                                                    const State* alias,
                                                    const Raw* raw,
                                                    const ManualTagged* tagged) {
    asm volatile("" : : "r"(state), "r"(alias), "r"(raw), "r"(tagged) : "memory");
    return *state == State::Negative && *alias == State::Alias && raw->integer == 42 &&
           tagged->kind == ManualTagged::Kind::Integer && tagged->payload.integer == 42;
}

int main() {
    State state = State::Negative;
    State alias = State::Alias;
    Raw raw{.integer = 42};
    ManualTagged tagged{
        .kind = ManualTagged::Kind::Integer,
        .payload = {.integer = 42},
    };
    return inspect_enums(&state, &alias, &raw, &tagged) ? 0 : 1;
}
